### PR TITLE
Optimize performance of substr_index and add tests

### LINKED ISSUE #9890: Optimize performance of the substr_index function
### Is your feature request related to a problem or challenge?

As noted in https://github.com/apache/arrow-datafusion/pull/9864#discussion_r1545353414 there might be some optimization available for the substr_index function so to facilitate that add a benchmark for that function.

A new benchmark testing the substr_index function was added in https://github.com/apache/arrow-datafusion/pull/9878 and available for use to test the performance of that function and any changes.

### Describe the solution you'd like

Make the `substr_index` benchmark faster

To run:
```shell
cargo bench --bench substr_index
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by kevinmingtarja**
## Which issue does this PR close?

Closes #9890 

## Rationale for this change

We were using dynamic dispatch when deciding whether to do `string.split()` or `string.rsplit()` (depending on the `count` argument) because the exact type of iterator is not known at compile time, plus it allocates memory on the heap for the boxed iterator, which introduces considerable overhead. 

Instead, we can avoid doing this and handle the two cases with an if else statement, which removes that overhead and potentially enable further optimizations by the compiler.

Another major overhead is pointed to in https://github.com/apache/arrow-datafusion/pull/9973#discussion_r1554566944. Instead, we can avoid `String` creation at every iteration of `map` by using `StringBuilder` to build the output.

### Benchmarks
```bash
substr_index_array_array_1000
                        time:   [57.082 µs 57.407 µs 57.770 µs]
                        change: [-53.118% -52.034% -51.077%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe
```

No more time wasted on `malloc` and `free` now if we see the [flamegraph](https://github.com/apache/arrow-datafusion/assets/69668484/e5a847da-4597-42ab-9dc6-45cc58bf741d):
Screen Shot 2024-04-07 at 03 36 49

## Are these changes tested?

Yes. I added new unit tests as well for `substr_index` since there were none before.

### CODE REVIEW

**Comment by kevinmingtarja** [comment]
CI [failure](https://github.com/apache/arrow-datafusion/actions/runs/8579761480/job/23515164131?pr=9973) seems unrelated:

```
error: failed to run `rustc` to learn about target-specific information

Caused by:
  process didn't exit successfully: `sccache /usr/local/rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc - --crate-name ___ --print=file-names -C debuginfo=line-tables-only -C incremental=false --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg` (exit status: 2)
  --- stderr
  sccache: error: Timed out waiting for server startup. Maybe the remote service is unreachable?
  Run with SCCACHE_LOG=debug SCCACHE_NO_DAEMON=1 to get more information
Error: Process completed with exit code 101.
```

**Comment by alamb** [comment]
> CI [failure](https://github.com/apache/arrow-datafusion/actions/runs/8579761480/job/23515164131?pr=9973) seems unrelated:

Agreed -- I restarted the job

File: datafusion/functions/src/unicode/substrindex.rs [hunk 1]
```diff
@@ -112,18 +112,22 @@ pub fn substr_index<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
...
                 };
-                let occurrences = usize::try_from(n.unsigned_abs()).unwrap_or(usize::MAX);
-                // The length of the substring covered by substr_index.
-                let length = splitted
-                    .take(occurrences) // at least 1 element, since n != 0
-                    .map(|s| s.len() + delimiter.len())
-                    .sum::<usize>()
-                    - delimiter.len();
                 if n > 0 {
                     Some(string[..length].to_owned())
```
Comment (alamb) [inline datafusion/functions/src/unicode/substrindex.rs:132]: This code looks much nicer -- thank you @kevinmingtarja 

Something else you could do is avoid the `String` creation here entirely -- creating `String`s for each row is non ideal for two reasons:
1. It does an allocation 
2. It needs to be copied *again* to the final output

Specifically, using https://docs.rs/arrow/latest/arrow/array/type.StringBuilder.html you could build up the output directly 

So something like

```rust
let mut builder = StringBuilder::new();
...
    if n > 0 {
      builder.append_value(string[..length]));
    } else {
      builder.append_value(string[string.len() - length..])
    }
...
```

The key is to avoid the `to_owned()` call

**Review by alamb — DISMISSED** [review]
Looks nice a nice improvement to me -- thank you @kevinmingtarja 🙏 

I left a suggestion for how to make it even better, but I also think we can do that as a follow on PR

File: datafusion/functions/src/unicode/substrindex.rs (hunk 1 above)
Comment (kevinmingtarja) [reply datafusion/functions/src/unicode/substrindex.rs:132]: Actually, since we haven't merged this, let me do this real quick!

And thanks for pointing me to this, I wasn't aware of Arrow's StringBuilder API before.

Comment (kevinmingtarja) [reply datafusion/functions/src/unicode/substrindex.rs]: Awesome, managed to achieve 50% speedup now in total!

```bash
substr_index_array_array_1000
                        time:   [57.082 µs 57.407 µs 57.770 µs]
                        change: [-53.118% -52.034% -51.077%] (p = 0.00 < 0.05)
                        Performance has improved.
```

I've just pushed the new commit.

File: datafusion/functions/src/unicode/substrindex.rs [hunk 2]
```diff
@@ -101,38 +101,151 @@ pub fn substr_index<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
...
-                    .sum::<usize>()
-                    - delimiter.len();
                 if n > 0 {
-                    Some(string[..length].to_owned())
+                    if let Some(substring) = string.get(..length) {
+                        builder.append_value(substring);
+                    }
                 } else {
-                    Some(string[string.len() - length..].to_owned())
+                    #[allow(clippy::collapsible_else_if)]
```
Comment (kevinmingtarja) [inline datafusion/functions/src/unicode/substrindex.rs:138]: I felt like collapsing this to `else if let Some(...) = ...` makes it less readable, which is why I override it.

File: datafusion/functions/src/unicode/substrindex.rs [hunk 3]
```diff
@@ -101,38 +101,154 @@ pub fn substr_index<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
...
-                let occurrences = usize::try_from(n.unsigned_abs()).unwrap_or(usize::MAX);
-                // The length of the substring covered by substr_index.
-                let length = splitted
-                    .take(occurrences) // at least 1 element, since n != 0
-                    .map(|s| s.len() + delimiter.len())
-                    .sum::<usize>()
-                    - delimiter.len();
                 if n > 0 {
-                    Some(string[..length].to_owned())
+                    if let Some(substring) = string.get(..length) {
```
Comment (alamb) [inline datafusion/functions/src/unicode/substrindex.rs:134]: For each row here, I think you either need to call `append_value` or `append_null` -- the code as I read it will not call `append_null` for non matching rows. I'll suggest some more tests below

File: datafusion/functions/src/unicode/substrindex.rs [hunk 4]
```diff
@@ -101,38 +101,154 @@ pub fn substr_index<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+                    }
                 }
             }
-            _ => None,
-        })
-        .collect::<GenericStringArray<T>>();
+        });
+
+    if builder.is_empty() {
+        builder.append_null();
```
Comment (alamb) [inline datafusion/functions/src/unicode/substrindex.rs:149]: I think append null needs to be called on each row (not just once at the end) -- this will work fine for tests that only use a single row array (like these unit tests) but likely won't work for arrays with multiple rows 


It might be useful to add some tests here for NULL inputs :

https://github.com/apache/arrow-datafusion/blob/1a002bccd420ff91ec149ee1ba9c42061510f906/datafusion/sqllogictest/test_files/functions.slt#L939-L954

**Review by alamb — COMMENTED** [review]
Thank you @kevinmingtarja  -- this is looking really close. I think the null/no match cases aren't quite right but it should be easily fixable

File: datafusion/functions/src/unicode/substrindex.rs (hunk 3 above)
Comment (kevinmingtarja) [reply datafusion/functions/src/unicode/substrindex.rs]: Ah I see, got it. I've made a fix a for this.

File: datafusion/functions/src/unicode/substrindex.rs (hunk 4 above)
Comment (kevinmingtarja) [reply datafusion/functions/src/unicode/substrindex.rs]: Got it, I've added a new test in that file.

File: datafusion/functions/src/unicode/substrindex.rs [hunk 5]
```diff
@@ -101,38 +101,151 @@ pub fn substr_index<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+                    }
                 }
             }
-            _ => None,
-        })
-        .collect::<GenericStringArray<T>>();
+            _ => builder.append_null(),
+        });
+
+    Ok(Arc::new(builder.finish()) as ArrayRef)
```
Comment (alamb) [inline datafusion/functions/src/unicode/substrindex.rs:148]: 👨‍🍳 👌   -- very nice

File: datafusion/sqllogictest/test_files/functions.slt [hunk 6]
```diff
@@ -954,6 +955,14 @@ SELECT str, n, substring_index(str, '.', n) AS c FROM
   ) AS occurrences(n)
 ORDER BY str DESC, n;
 ----
+NULL -100 NULL
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/functions.slt:958]: 💯  thank you

**Review by alamb — APPROVED** [review]
Thanks @kevinmingtarja  -- this looks really nice 🙏

**Comment by alamb** [comment]
FYI @Omega359

File: datafusion/functions/src/unicode/substrindex.rs [hunk 7]
```diff
@@ -101,38 +101,151 @@ pub fn substr_index<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+                    builder.append_value("");
+                    return;
                 }
 
-                let splitted: Box<dyn Iterator<Item = _>> = if n > 0 {
-                    Box::new(string.split(delimiter))
+                let occurrences = usize::try_from(n.unsigned_abs()).unwrap_or(usize::MAX);
+                let length = if n > 0 {
+                    let splitted = string.split(delimiter);
+                    splitted
```
Comment (comphead) [inline datafusion/functions/src/unicode/substrindex.rs:120]: Nice, I would factor our this code into internal function

Comment (comphead) [reply datafusion/functions/src/unicode/substrindex.rs:120]: can be done in follow up PR @kevinmingtarja

**Review by comphead — APPROVED** [review]
lgtm thanks @kevinmingtarja I really like the improvement

**Comment by Omega359** [comment]
Nice work! If there is another function that you think could use a look at for performance I'll be happy to write the benchmark for testing.

File: datafusion/functions/src/unicode/substrindex.rs (hunk 7 above)
Comment (kevinmingtarja) [reply datafusion/functions/src/unicode/substrindex.rs:120]: got it

**Comment by kevinmingtarja** [comment]
Thanks for the help and review @alamb, @comphead, and thanks @Omega359 for writing the benchmark!

**Comment by alamb** [comment]
🚀
